packages/mdc-checkbox/component.ts

```diff
@@ -0,0 +1,183 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+import {getCorrectEventName} from '@material/animation/util';
+import {MDCComponent} from '@material/base/component';
+import {EventType, SpecificEventListener} from '@material/base/types';
+import {ponyfill} from '@material/dom/index';
+import {MDCRipple, MDCRippleFoundation, RippleCapableSurface} from '@material/ripple/index';
+import {MDCSelectionControl} from '@material/selection-control/index';
+import {MDCCheckboxFoundation} from './foundation';
+
+const CB_PROTO_PROPS = ['checked', 'indeterminate'];
+
+export type MDCCheckboxFactory = (root: Element, foundation?: MDCCheckboxFoundation) => MDCCheckbox;
+
+export class MDCCheckbox extends MDCComponent<MDCCheckboxFoundation>
+  implements MDCSelectionControl, RippleCapableSurface {
+
+  static attachTo(root: Element) {
+    return new MDCCheckbox(root);
+  }
+
+  /**
+   * Returns the state of the native control element, or null if the native control element is not present.
+   */
+  get nativeCb_(): HTMLInputElement {
+    const {NATIVE_CONTROL_SELECTOR} = MDCCheckboxFoundation.strings;
+    const cbEl = this.root_.querySelector<HTMLInputElement>(NATIVE_CONTROL_SELECTOR);
+    if (!cbEl) {
+      throw new Error(`Checkbox requires a ${NATIVE_CONTROL_SELECTOR} element`);
+    }
+    return cbEl;
+  }
+
+  // Public visibility for this property is required by RippleCapableSurface.
+  root_!: Element; // assigned in MDCComponent constructor
+
+  private ripple_: MDCRipple = this.initRipple_();
+  private handleChange_!: EventListener; // assigned in initialSyncWithDOM()
+  private handleAnimationEnd_!: EventListener; // assigned in initialSyncWithDOM()
+
+  initialSyncWithDOM() {
+    this.handleChange_ = () => this.foundation_.handleChange();
+    this.handleAnimationEnd_ = () => this.foundation_.handleAnimationEnd();
+    this.nativeCb_.addEventListener('change', this.handleChange_);
+    this.listen(getCorrectEventName(window, 'animationend'), this.handleAnimationEnd_);
+    this.installPropertyChangeHooks_();
+  }
+
+  destroy() {
+    this.ripple_.destroy();
+    this.nativeCb_.removeEventListener('change', this.handleChange_);
+    this.unlisten(getCorrectEventName(window, 'animationend'), this.handleAnimationEnd_);
+    this.uninstallPropertyChangeHooks_();
+    super.destroy();
+  }
+
+  getDefaultFoundation(): MDCCheckboxFoundation {
+    return new MDCCheckboxFoundation({
```

Comment:
Extract variable for all adapter implementations.

The foundation argument is a `Partial<MDCFooAdapter>`, so `tsc` won't complain if we omit a method; it'll just merge it with the default implementation, which does nothing.

Extracting a variable and explicitly declaring it of type `MDCFooAdapter` ensures that we'll get a compiler error if we forget a field.